DOC: Start to document interactive figures (#4779)
matplotlib/matplotlib: doc/users/interactive.rst

can be an expensive operation, and you do not want to re-render the
figure every time a single property is changed, only once after all
the properties have changed and the figure is displayed on the screen
or saved to disk.  When working in a shell, one typically wants the
figure to re-draw after every user command (ex when the prompt comes
back).  *interactive* mode of :mod:`~.pyplot` takes care of arranging
such that if any open figures are :ref:`stale <stale_artists>`, they
will be re-drawn just before the prompt is returned to the user.

.. _ipython-pylab:

IPython to the rescue
=====================

We recommend using IPython for an interactive shell.  In addition to
all of it's features (improved tab-completion, magics,
multiline-editing, etc), it also ensures that the GUI toolkit is
properly integrated with the command line (see :ref:`cp_integration`).
To configure the integration and enable interactive mode do::

.. sourcecode:: ipython

   user@machine:~ $ ipython
   Python 3.6.4 (default, Dec 23 2017, 19:07:07)
   Type 'copyright', 'credits' or 'license' for more information
   IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

   In [1]: %matplotlib

>    iUsing matplotlib backend: Qt5Agg

Reviewer: extra "i"

.. highlight:: python

.. _controlling-interactive:

Interactive mode
================

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

   pyplot.ion
   pyplot.ioff
   pyplot.isinteractive

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

   pyplot.show
   pyplot.pause

Interactive mode controls:

- whether created figures are automatically shown
- whether changes to artists automatically trigger re-drawing existing figures
- whether `.pyplot.show` blocks or not

If in interactive mode, then:

- newly created figures will be shown immediately
- figures will automatically redraw on change
- pyplot.show will not block by default

If not in interactive mode then:

- newly created figures and changes to figures will
  not be reflected until explicitly asked to be
- pyplot.show will block by default

If you are in non-interactive mode (or created figures while in
non-interactive mode) you may need to explicitly call `.pyplot.show`
to bring the windows onto your screen.  If you only want to run the
GUI event loop for a fixed amount of time you can use `.pyplot.pause`.

Being in interactive mode is orthogonal to the GUI event loop being

> integrated with your command prompt.  If you have the GUI event loop

Author: No, this is the right way around.  The event loop for the prompt runs the show, it in turn spins the GUI event loop while waiting for the slow humans to type.

integrated with your prompt, then shown figures will be "live" while
the prompt is waiting for input, if it is not integrated then your

> figures will only be "live" when the GUI event loop is running (via

Author: At least PyQt and tkinter have ship with hooks that integrate with the readline implementation of the "plain" Python prompt.  The IPython magic is the easiest way but not the only way.

   will likely result in a non-responsive figure.

.. _navigation-toolbar:

Default UI
==========

The windows created by :mod:`~.pyplot` have an interactive toolbar with navigation
buttons and a readout of where the cursor is in dataspace.  A number of
helpful keybindings are registered by default.

.. _key-event-handling:

Navigation Keyboard Shortcuts
-----------------------------

The following table holds all the default keys, which can be
overwritten by use of your matplotlibrc (#keymap.\*).

================================== =================================================
Command                            Keyboard Shortcut(s)
================================== =================================================
Home/Reset                         **h** or **r** or **home**
Back                               **c** or **left arrow** or **backspace**
Forward                            **v** or **right arrow**
Pan/Zoom                           **p**
Zoom-to-rect                       **o**
Save                               **ctrl** + **s**
Toggle fullscreen                  **f** or **ctrl** + **f**
Close plot                         **ctrl** + **w**

> Close all plots                    **shift** + **w**

Author: https://github.com/matplotlib/matplotlib/blob/7b7017eb53248438ff070ccb135b7eda9a2974ba/lib/matplotlib/backend_bases.py#L2377-L2378

It looks like it is still there?  Is there a PR for that in the wings?

> Close all plots                    **shift** + **w**

Reviewer 2: There are no longer any default shortcuts #16863

Zoom-to-rect                       :rc:`keymap.zoom`
Save                               :rc:`keymap.save`
Toggle fullscreen                  :rc:`keymap.fullscreen`
Toggle major grids                 :rc:`keymap.grid`
Toggle minor grids                 :rc:`keymap.grid_minor`
Toggle x axis scale (log/linear)   :rc:`keymap.xscale`
Toggle y axis scale (log/linear)   :rc:`keymap.yscale`
Close Figure                       :rc:`keymap.quit`
Constrain pan/zoom to x axis       hold **x** when panning/zooming with mouse
Constrain pan/zoom to y axis       hold **y** when panning/zooming with mouse

> Preserve aspect ratio              hold **CONTROL** when panning/zooming with mouse

Author: right-click pan will change the view limits and respect this.

> Preserve aspect ratio              hold **CONTROL** when panning/zooming with mouse

Reviewer 3: I've never heard of right-click pan before, as far as I know.  Maybe I've never stumbled over it because I use the Mac trackpad rather than a mouse.

> Preserve aspect ratio              hold **CONTROL** when panning/zooming with mouse

Author: right-click zoom zooms out (by putting the current view limits to be where the box you drew was)

   >>> plt.ion()
   >>>

however this does not ensure that the event hook is properly installed
and your figures may not be responsive.  Please consult the
documentation of your GUI toolkit for details.

Jupyter Notebooks / Lab
-----------------------

.. note::

   To get the interactive functionality described here, you must be
   using an interactive backend.  The default backend in notebooks,
   the inline backend, is not.  `~ipykernel.pylab.backend_inline`
   renders the figure once and inserts a static image into the
   notebook when the cell is executed.  The images are static and can
   not be panned / zoomed, take user input, or be updated from other
   cells.

To get interactive figures in the 'classic' notebook or jupyter lab
use the `ipympl <https://github.com/matplotlib/ipympl>`__ backend
(must be installed separately) which uses the **ipywidget** framework.
If ``ipympl`` is installed use the magic:

.. sourcecode:: ipython

  %matplotlib widget

to select and enable it.

If you only need to use the classic notebook you can use

.. sourcecode:: ipython

  %matplotlib notebook

which uses the `.backend_nbagg` backend which ships with Matplotlib.

> However nbagg does not work in Jupyter Lab due to changes in the front

Reviewer 3: Maybe delete "due to changes...".  It's not clear what is meant by "the front end".

end.

GUIs + jupyter
~~~~~~~~~~~~~~

If you are running your jupyter kernel locally you can use one of the
GUI backends.  The process running your kernel will show a GUI window
on your desktop adjacent to your web browser.  However if you move
that notebook to a remote server the kernel will try to open the GUI
window on *that* computer.  Unless you have arranged to forward the
xserver back to your desktop, you not be able to see or interact with
the figure (if it does not raise an exception outright).

PyCharm, Spyder, and VSCode
---------------------------

> Many IDEs have built-in integration with Matplotlib, please consult its

Reviewer 3: "its" -> "their", and change the comma to a semicolon

.. currentmodule:: matplotlib

.. _mpl-shell:

=====================
 Interactive Figures
=====================

.. toctree::

When working with data it is often invaluable to be able to interact
with your plots. In many cases the built in pan/zoom and mouse-location
tools are sufficient, but you can also use the Matplotlib event system
to build customized data exploration tools.

Matplotlib ships with :ref:`backends <what-is-a-backend>` binding to
several GUI toolkits (Qt, Tk, Wx, Gtk, OSX, js) and third party
packages provide bindings to `kivy

> <https://github.com/kivy-garden/garden.matplotlib>`__ and `Jupyter Lab

Author: Last scipy I helped someone who was actively using it (to built a pretty cool tool for looking at bird song audio!).  I would prefer to keep linking it unless we know it is broken.

   Python 3.8.2 (default, Apr  8 2020, 14:31:25)
   Type 'copyright', 'credits' or 'license' for more information
   IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

   In [1]: %matplotlib
   Using matplotlib backend: Qt5Agg

   In [2]: import matplotlib.pyplot as plt

   In [3]:

Calling

::

   In [3]: fig, ax = plt.subplots()

will pop open a window for you and

::

   In [4]: ln, = ax.plot(range(5))

will show your data in the window.  If you change something about the
line, for example the color

::

   In [5]: ln.set_color('orange')

it will be reflected immediately. If you wish to disable this behavior
use

::

   In [6]: plt.ioff()

and

::

   In [7]: plt.ion()

re-enable it.

With recent versions of ``Matplotlib`` and ``IPython`` it is
sufficient to import `matplotlib.pyplot` and call `.pyplot.ion`, but
using the magic is guaranteed to work regardless of versions.

.. highlight:: python

.. _controlling-interactive:

Interactive mode
================

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

   pyplot.ion
   pyplot.ioff
   pyplot.isinteractive

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

>    pyplot.show

Author: I was trying to group `ion`, `ioff`, `isinteractive` as a "family" that queries and modifies the state of interactive mode and these two as a "family" that may run the GUI event loop.  I'm not sure if it works as well in practice as it does in my head.

   the inline backend, is not.  `~ipykernel.pylab.backend_inline`
   renders the figure once and inserts a static image into the
   notebook when the cell is executed.  The images are static and can
   not be panned / zoomed, take user input, or be updated from other
   cells.

To get interactive figures in the 'classic' notebook or jupyter lab
use the `ipympl <https://github.com/matplotlib/ipympl>`__ backend
(must be installed separately) which uses the **ipywidget** framework.
If ``ipympl`` is installed use the magic:

.. sourcecode:: ipython

  %matplotlib widget

to select and enable it.

If you only need to use the classic notebook you can use

.. sourcecode:: ipython

  %matplotlib notebook

which uses the `.backend_nbagg` backend which ships with Matplotlib.
However nbagg does not work in Jupyter Lab due to changes in the front
end.

GUIs + jupyter
~~~~~~~~~~~~~~

If you are running your jupyter kernel locally you can use one of the
GUI backends.  The process running your kernel will show a GUI window
on your desktop adjacent to your web browser.  However if you move
that notebook to a remote server the kernel will try to open the GUI

> window on _that_ computer.  Unless you have arranged to forward the

Reviewer 2: ```suggestion
window on *that* computer.  Unless you have arranged to forward the
```

with your plots. In many cases the built in pan/zoom and mouse-location
tools are sufficient, but you can also use the Matplotlib event system
to build customized data exploration tools.

Matplotlib ships with :ref:`backends <what-is-a-backend>` binding to
several GUI toolkits (Qt, Tk, Wx, Gtk, OSX, js) and third party
packages provide bindings to `kivy
<https://github.com/kivy-garden/garden.matplotlib>`__ and `Jupyter Lab
<https://github.com/matplotlib/ipympl>`__.  For the figures to be
"live" the GUI event loop will need to be integrated with your prompt. The
simplest way is to use IPython (see :ref:`below <ipython-pylab>`).

The `.pyplot` module provides functions for explicitly creating
Figures that include a interactive tools, a toolbar, a tool-tip, and
:ref:`key bindings <key-event-handling>` ready to go:

`.pyplot.figure`
    Creates a new empty `.figure.Figure` or selects an existing figure

`.pyplot.subplots`
    Creates a new `.figure.Figure` and fills it with a grid of `.axes.Axes`

`.pyplot` has a notion of "The Current Figure" which can be accessed
through `.pyplot.gcf` and a notion of "The Current Axes" accessed

> through `.pyplot.gca`.  Almost all of the function is `.pyplot` as pass
> through to the current `.Figure` / `.axes.Axes` (or create one) as
> appropriate.  Matplotlib keeps a reference to all of the open figures

Reviewer 2: ```suggestion
through `.pyplot.gca`.  Almost all of the functions in `.pyplot` pass
through the current `.Figure` / `.axes.Axes` (or create one) as
appropriate.  Matplotlib keeps a reference to all of the open figures
```

buttons and a readout of where the cursor is in dataspace.  A number of
helpful keybindings are registered by default.

.. _key-event-handling:

Navigation Keyboard Shortcuts
-----------------------------

The following table holds all the default keys, which can be
overwritten by use of your :ref:`matplotlibrc
<sphx_glr_tutorials_introductory_customizing.py>`.

================================== ===============================
Command                            Default key binding and rcParam
================================== ===============================
Home/Reset                         :rc:`keymap.home`
Back                               :rc:`keymap.back`
Forward                            :rc:`keymap.forward`
Pan/Zoom                           :rc:`keymap.pan`
Zoom-to-rect                       :rc:`keymap.zoom`
Save                               :rc:`keymap.save`
Toggle fullscreen                  :rc:`keymap.fullscreen`
Toggle major grids                 :rc:`keymap.grid`
Toggle minor grids                 :rc:`keymap.grid_minor`
Toggle x axis scale (log/linear)   :rc:`keymap.xscale`
Toggle y axis scale (log/linear)   :rc:`keymap.yscale`

> Toggle y axis scale (log/linear)   :rc:`keymap.yscale`

Reviewer 2: Double entry.

documentation of your GUI toolkit for details.

Jupyter Notebooks / Lab
-----------------------

.. note::

   To get the interactive functionality described here, you must be
   using an interactive backend.  The default backend in notebooks,
   the inline backend, is not.  `~ipykernel.pylab.backend_inline`
   renders the figure once and inserts a static image into the
   notebook when the cell is executed.  The images are static and can
   not be panned / zoomed, take user input, or be updated from other
   cells.

To get interactive figures in the 'classic' notebook or jupyter lab
use the `ipympl <https://github.com/matplotlib/ipympl>`__ backend
(must be installed separately) which uses the **ipywidget** framework.
If ``ipympl`` is installed use the magic:

.. sourcecode:: ipython

  %matplotlib widget

to select and enable it.

If you only need to use the classic notebook you can use

.. sourcecode:: ipython

  %matplotlib notebook

which uses the `.backend_nbagg` backend which ships with Matplotlib.
However nbagg does not work in Jupyter Lab due to changes in the front
end.

GUIs + jupyter
~~~~~~~~~~~~~~

If you are running your jupyter kernel locally you can use one of the

> GUI backends.  Process running your kernel will and shows a GUI window
> on your desktop adjacent to your web browser.  However if you move

Reviewer 2: ```suggestion
GUI backends.  The process running your kernel will show a GUI window
on your desktop adjacent to your web browser.  However if you move
```

that notebook to a remote server the kernel will try to open the GUI

> window on _that_ computer.  Unless you have arrange to forward the

Reviewer 2: ```suggestion
window on _that_ computer.  Unless you have arranged to forward the
```

The `.pyplot` module provides two functions for creating Figures that,
when an interactive backend is used, that are managed by Matplotlib,
embedded in GUI windows, and ready for interactive use out of the box

`.pyplot.figure`
    Creates a new empty `.figure.Figure` or selects an existing figure

`.pyplot.subplots`
    Creates a new `.figure.Figure` and fills it with a grid of `.axes.Axes`

Matplotlib keeps a reference to all of the open figures created this
way so they will not be garbage collected.  You can close all off your
open figures via ``plt.close('all')``.

For discussion of how this works under the hood see:

.. toctree::
    :maxdepth: 1

    interactive_guide.rst
    event_handling.rst

.. _ipython-pylab:

IPython integration
===================

We recommend using IPython for an interactive shell.  In addition to
all of its features (improved tab-completion, magics,
multiline-editing, etc), it also ensures that the GUI toolkit event
loop is properly integrated with the command line (see
:ref:`cp_integration`).  To configure the integration and enable
:ref:`interactive mode <controlling-interactive>` use the
``%matplotlib`` magic

> .. sourcecode:: ipython

Reviewer: instead of repeating `.. sourcecode:: ipython` you can probably use https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-highlight

   user@machine:~ $ ipython
   Python 3.8.2 (default, Apr  8 2020, 14:31:25)
   Type 'copyright', 'credits' or 'license' for more information
   IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

   In [1]: %matplotlib
   Using matplotlib backend: Qt5Agg

   In [2]: import matplotlib.pyplot as plt

   In [3]:

Calling

.. sourcecode:: ipython

   In [3]: fig, ax = plt.subplots()

will pop open a window for you and

.. sourcecode:: ipython

   In [4]: ln, = ax.plot(range(5))

will show your data in the window.  If you

> want to change anything about the line, ex the color

Reviewer: "ex the color" typo

Author: change this to "for example the color", but not seeing the typo.

Forward                            **v** or **right arrow**
Pan/Zoom                           **p**
Zoom-to-rect                       **o**
Save                               **ctrl** + **s**
Toggle fullscreen                  **f** or **ctrl** + **f**
Close plot                         **ctrl** + **w**
Close all plots                    **shift** + **w**
Constrain pan/zoom to x axis       hold **x** when panning/zooming with mouse
Constrain pan/zoom to y axis       hold **y** when panning/zooming with mouse
Preserve aspect ratio              hold **CONTROL** when panning/zooming with mouse
Toggle major grids                 **g** when mouse is over an axes
Toggle minor grids                 **G** when mouse is over an axes
Toggle x axis scale (log/linear)   **L** or **k**  when mouse is over an axes
Toggle y axis scale (log/linear)   **l** when mouse is over an axes
================================== =================================================

.. _other-shells:

Other Python prompts
====================

If you can not or do not want to use IPython, interactive mode works
in the vanilla python prompt

> .. sourcecode:: python

Reviewer: "pycon", I guess https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer

   >>> import matplotlib.pyplot as plt
   >>> plt.ion()
   >>>

however this does not ensure that the event hook is properly installed
and your figures may not be responsive.  Please consult the
documentation of your GUI toolkit for details.

Jupyter Notebooks / Lab
-----------------------

.. note::

   To get the interactive functionality described here, you must be
   using an interactive backend.  The default backend in notebooks,
   the inline backend, is not.  `~ipykernel.pylab.backend_inline`
   renders the figure once and inserts a static image into the
   notebook when the cell is executed.  The images are static and can
   not be panned / zoomed, take user input, or be updated from other
   cells.

Jupyter uses a different architecture than a traditional interactive
terminal.  Rather than the user interface running in the same process
as your interpreter, the user interacts with a javascript front end
running in a browser which communicates with a server which in turn
communicates with a kernel that executes your Python.  This means
for interactive figures our UI needs to also be written in javascript

> and run inside of the jupyter front end.

Reviewer: may be too technical for the average user?  (could be mentioned later)

(must be installed separately) which uses the **ipywidget** framework.
If ``ipympl`` is installed use the magic:

.. sourcecode:: ipython

  %matplotlib widget

to select and enable it.

If you only need to use the classic notebook you can use

.. sourcecode:: ipython

  %matplotlib notebook

which uses the `.backend_nbagg` backend which ships with Matplotlib.
However nbagg does not work in Jupyter Lab due to changes in the front
end.

GUIs + jupyter
~~~~~~~~~~~~~~

If you are running your jupyter server locally you can use one of the
GUI backends.  However if you ever move that notebook to a remote
server it will cease to work correctly because the GUI windows will be
created on the server, not your machine. When you create a figure the
process running your kernel creates and shows a GUI window.  If that
process is on the same computer as your client, then you will be able
to see and interact with the window.  However if it is running on a
remote computer it will try to open the GUI window on _that_ computer.
This will either fail by raising an exception (as many servers
do not have an XServer running) or run, but leave you with no
way to access your figure.

> PyCharm, Spydter, and VSCode

Reviewer: spyder typo

   pyplot.ion
   pyplot.ioff
   pyplot.isinteractive

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

   pyplot.show
   pyplot.pause

Interactive mode controls:

- whether created figures are automatically shown
- whether changes to artists automatically trigger re-drawing existing figures
- whether `.pyplot.show` returns immediately or after all of the
  figures have been closed when given no arguments

If in interactive mode, then:

- newly created figures will be shown immediately
- figures will automatically redraw on change
- pyplot.show will return immediately by default

If not in interactive mode then:

- newly created figures and changes to figures will
  not be reflected until explicitly asked to be
- pyplot.show runs the GUI event loop and does not return until all of
  the plot windows are closed

If you are in non-interactive mode (or created figures while in
non-interactive mode) you may need to explicitly call `.pyplot.show`
to bring the windows onto your screen.  If you only want to run the
GUI event loop for a fixed amount of time you can use `.pyplot.pause`.
This will both block the progress of your code (as if you had called

> `time.sleep`), raise ensure the current window is shown and if needed

Reviewer 2: ```suggestion
`time.sleep`), ensure the current window is shown and if needed
```

re-drawn, and run the GUI event loop (so the windows are "live" for

> intrection) for the specified period of time.

Reviewer 2: ```suggestion
interaction) for the specified period of time.
```

Being in "interactive mode" is orthogonal to the GUI event loop being
integrated with your command prompt.  If you use `pyplot.ion`, but

> have not arrange for the event loop integration your figures will

Reviewer 2: ```suggestion
have not arranged for the event loop integration, your figures will
```

appear but will not be "live" while the prompt is waiting for input.
You will not be able to pan/zoom and the figure may not even render

> (the window might appear black, transparent, or as a snapshot the

Reviewer 2: ```suggestion
(the window might appear black, transparent, or as a snapshot of the
```

desktop under it).  Conversely, if you configure the event loop

> integration displayed figures will be "live" while waiting for input
> at the prompt independent of pyplot's "interactive mode".  In either

Reviewer 2: ```suggestion
integration, displayed figures will be "live" while waiting for input
at the prompt, regardless of pyplot's "interactive mode".  In either
```

just before the figure is displayed on the screen or saved to disk.
In the case of scripts this is easy to arrange, we simple do nothing
until ``savefig`` is called.  However, when working interactively,
either at a terminal or in a notebook, we would like the figure to
automatically re-render when we change.  Further, we would like the
figures respond to user input (mouse and keyboard) to interact with
and explore the data.

To get this functionality we need a combination of an interactive
backend (to handle the user interaction with the figure) and an
interactive interpreter (to handle the input of user code).  Together
these tools need to:

- Take text from the user and execute it in the Python interpreter
- Ensure that any created figures are put on the screen
- Trigger a re-render of the figure when the user has mutated it
- When not actively executing user input, run the event loop so we can
  get user input on the figure (see
  :ref:`interactive_figures_and_eventloops` for details)

.. _ipython-pylab:

IPython to the rescue!
======================

To support interactive

We recommend using IPython for an interactive shell.  In addition to

> all of it's features (improved tab-completion, magics,

Reviewer 4: it's -> its

   In [1]: %matplotlib
   iUsing matplotlib backend: Qt5Agg

   In [2]: import matplotlib.pyplot as plt

   In [3]:

Calling

.. sourcecode:: ipython

   In [3]: fig, ax = plt.subplots()

will pop open a window for you and

.. sourcecode:: ipython

   In [4]: ln, = ax.plot(range(5))

will show your data in the window.  If you
want to change anything about the line, ex the color

.. sourcecode:: ipython

   In [5]: ln.set_color('orange')

will be reflected immediately.

.. _other-shells:

Other Python interpreters
=========================

If you can not or do not want to use IPython, interactive mode
should work in the vanilla python prompt

.. sourcecode:: python

   user@machine:~ $ python
   Python 3.6.4 (default, Dec 23 2017, 19:07:07)
   [GCC 7.2.1 20171128] on linux
   Type "help", "copyright", "credits" or "license" for more information.
   >>> import matplotlib.pyplot as plt
   >>> plt.ion()
   >>>

however this does not ensure that the event hook is properly installed
and your figures may not be responsive (see :ref:`cp_integration`).

GUI python prompts may or may not work with all backends.  For
example, if you use the IDLE IDE you must use the 'tkagg' backend (as

> IDLE is a Tk application..  However, if you use spyder, the

Reviewer 5: extra .

interactive shell is run in a sub process of the GUI and you can use
any backend.

PyCharm
-------

TODO: links to pycharm docs on configuring Matplotlib backends

Spyder
------

TODO: links to spyder docs on configuring Matplotlib backends

VSCode
------

TODO: links to vscode docs on configuring Matplotlib backends (?)

Jupyter Notebooks / Jupyter Lab
-------------------------------

.. warning::

   To get the interactive functionality described here, you must be

>    using an interactive backend, however he 'inline' backend is not.

Reviewer 5: s/he/the/

fail by raising an exception (as many linux servers do not have an
XServer running) or run cleanly, but leave you with no way to access
your figure.

.. _controlling-interactive:

Controlling interactive updating
================================

To control and query the current state of *interactive* mode

.. autosummary::
   :template: autosummary.rst
   :nosignatures:

   pyplot.ion
   pyplot.ioff
   pyplot.isinteractive

When working with a big figure in which drawing is expensive, you may
want to turn matplotlib's interactive setting off temporarily to avoid
the performance hit::

    >>> #create big-expensive-figure
    >>> plt.ioff()      # turn updates off
    >>> ax.set_title('now how much would you pay?')
    >>> fig.canvas.draw_idle()      # force a draw
    >>> fig.savefig('alldone', dpi=300)
    >>> plt.close('all')
    >>> plt.ion()      # turn updating back on
    >>> fig, ax = plt.subplots()
    >>> ax.plot(rand(20), mfc='g', mec='r', ms=40, mew=4, ls='--', lw=3)

Default UI
==========

.. toctree::
    :maxdepth: 1

    navigation_toolbar.rst

The windows created by :mod:`~.pyplot` have an interactive toolbar and

> has a number of helpful keybindings by default.

Reviewer 5: s/has/have/
